File: 101.py
# LeetCode 101 Symmetric Tree
# Reference: https://www.youtube.com/watch?v=Mao9uzxwvmc


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        def dfs(left, right):
            if not left and not right: # both nodes are null
                return True
            if not left or not right: # only one node is null
                return False
            
            return (left.val == right.val and dfs(left.left, right.right) and dfs(left.right, right.left))
        
        return dfs(root.left, root.right)  























        

File: 101.js
// LeetCode 101 Symmetric Tree
// Reference:https://leetcode.com/problems/symmetric-tree/solutions/494632/javascript-95-speed-100-memory-w-comments/?envType=study-plan-v2&envId=top-interview-150

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {

    if (root == null) return true;

    return symmetryChecker(root.left, root.right);
    
};

function symmetryChecker(left, right){
    // If both sub trees are empty
    if (left == null && right == null) return true;
    // If only one of the sub trees are empty
    if (left == null || right == null) return false;
    // If the values dont match up
    if (left.val != right.val) return false;

    // Check both subtrees but travelled in a mirrored/symmetric fashion
	// (one goes left, other goes right)  and make sure they're both symmetric

    return symmetryChecker(left.left, right.right) && symmetryChecker(left.right, right.left);
}